### PR TITLE
aggregate, reduce the complexity of the loop checking the overlapped fields

### PR DESCRIPTION
**Pull request by ghost**
Based on the fact that overlapping is only possible in unions and anonymous unions (yes ?) it's possible to avoid a nested control loop iff 

1. the aggregate is not an union. 
2. the aggregate doesn't contain any field declared in an anonymous union.

Little drawback : the second condition requires to add a field to the `VarDeclaration` class since `AnonDeclaration` (i.e anonymous union or anonymous struct) is implemented as an attribute.

### CODE REVIEW

File: src/dmd/declaration.d [hunk 1]
```diff
@@ -1080,6 +1080,7 @@ extern (C++) class VarDeclaration : Declaration
     bool mynew;                     // it is a class new'd with custom operator new
 
     int canassign;                  // it can be assigned to
+    bool anonUnionAttr;             // it was declared in an anon union
```
Comment (jacob-carlborg) [inline src/dmd/declaration.d]: All new public symbols should have a Ddoc comment. Please avoid abbreviations.

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/declaration.d [hunk 2]
```diff
@@ -1088,6 +1088,9 @@ extern (C++) class VarDeclaration : Declaration
     VarDeclaration lastVar;         // Linked list of variables for goto-skips-init detection
     uint endlinnum;                 // line number of end of scope that this var lives in
 
+    /// it was declared in an anon union
+    bool isAnonUnionMember;
```
Comment (ZombineDev) [inline src/dmd/declaration.d]: When changing the layout of `extern (C++)` aggregates, be use to update the header files as well.

Comment (Basile-z) [reply src/dmd/declaration.d]: yes fixup coming

Comment (Basile-z) [reply src/dmd/declaration.d]: https://github.com/dlang/dmd/pull/9290

**Comment by kinke** [comment]
AFAICT this led to a regression, some overlapping fields in the following example apparently don't get their `VarDeclaration.overlapped` set:
```D
struct SWithUnion
{
    char c;
    S nested;

    union
    {
        struct { ubyte ub = 6; ushort us = 33; align(8) ulong ul_dummy = void; ulong last = 123; }
        struct { uint ui1; uint ui2 = 84; ulong ul = 666; }
    }
}
```

**Comment by kinke** [comment]
> 2. the aggregate doesn't contain any field declared in an anonymous union

This goes more than one level deep, that's the problem, this code just looks at the direct variable children of anonymous unions. It would have to descend into nested anonymous unions and structs too.

I take it this was an attempt to improve performance (with any noticeable gains?); having to descend into subtrees etc. just makes this more complex and brittle, so I recommend a revert for 2.085.

**Comment by WalterBright** [comment]
@kinke can you turn this into a bugzilla issue, and add a piece of code that fails under this PR?

**Comment by kinke** [comment]
I'm swamped with LDC work.

**Comment by WalterBright** [comment]
I understand, but you do have a test case that triggers the problem, it just needs to be purified.

**Comment by WalterBright** [comment]
@kinke I created a bugzilla for this, https://issues.dlang.org/show_bug.cgi?id=19685 can you see about the test case?

**Comment by kinke** [comment]
Done, thx for bugzillaing.

**Comment by Basile-z** [comment]
Well, revert seems to be appropriated here since the commit was supposed to do the same thing.

**Comment by Basile-z** [comment]
The problem is that align was not handled at all by this PR, so again, revert.

**Comment by wilzbach** [comment]
Link to the revert PR: https://github.com/dlang/dmd/pull/9373

**Pull request merged**